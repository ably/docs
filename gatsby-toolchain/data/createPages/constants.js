const DEFAULT_LANGUAGE = 'default';
const TEXT_LANGUAGE = 'text';
const HYPERTEXT_LANGUAGE = 'html';
const YETANOTHERMARKUP_LANGUAGE = 'yaml';
const JSON_LANGUAGE = 'json';
const IGNORED_LANGUAGES = [DEFAULT_LANGUAGE, TEXT_LANGUAGE, HYPERTEXT_LANGUAGE, YETANOTHERMARKUP_LANGUAGE, JSON_LANGUAGE];
const IGNORED_LANGUAGES_FOR_DISPLAY = [TEXT_LANGUAGE, HYPERTEXT_LANGUAGE, YETANOTHERMARKUP_LANGUAGE, JSON_LANGUAGE];

module.exports = {
    DEFAULT_LANGUAGE,
    IGNORED_LANGUAGES,
    IGNORED_LANGUAGES_FOR_DISPLAY
};
