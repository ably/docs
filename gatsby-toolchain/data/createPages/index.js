const { flattenContentOrderedList, maybeRetrievePartial } = require("../transform");
const path = require("path");

const createPages = async ({ graphql, actions: { createPage } }) => {
    const documentTemplate = path.resolve(`src/templates/document.js`)
    const result = await graphql(`
      query {
        allFileHtml {
          edges {
            node {
              slug
              contentOrderedList {
                data
                type
              }
            }
          }
        }
      }
    `);
    const retrievePartialFromGraphQL = maybeRetrievePartial(graphql);
    await Promise.all(result.data.allFileHtml.edges.map(async edge => {
      const contentOrderedList = flattenContentOrderedList(await Promise.all(edge.node.contentOrderedList.map(retrievePartialFromGraphQL)));
  
      createPage({
        path: `/${edge.node.slug}`,
        component: documentTemplate,
        context: {
          slug: edge.node.slug,
          contentOrderedList
        },
      });
    }));
  }

module.exports ={ createPages };