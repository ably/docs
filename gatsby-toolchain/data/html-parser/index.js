const DataTypes = require("../types")

const htmlParser = content => {
    return [{
        data: content,
        type: DataTypes.Html
    }];
}

module.exports = {
    htmlParser
}