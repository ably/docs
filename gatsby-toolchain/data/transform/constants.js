const LATEST_ABLY_API_VERSION = 1.2;
const LATEST_ABLY_API_VERSION_STRING = '1.2';

module.exports = {
    LATEST_ABLY_API_VERSION,
    LATEST_ABLY_API_VERSION_STRING
};