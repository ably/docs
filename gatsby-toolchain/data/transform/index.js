const yaml = require('js-yaml');
const { upperFirst, camelCase, identity, isPlainObject, lowerFirst, isEmpty, merge } = require('lodash');
const { tryRetrieveMetaData, filterAllowedMetaFields, NO_MATCH } = require("./front-matter");
const DataTypes = require("../types");
const { ROOT_LEVEL, MAX_LEVEL } = require("../../src/components/Sidebar/consts");
const { preParser } = require("./pre-parser");
const { enhancedParse } = require("./parser-enhancements");

const INLINE_TOC_REGEX = /^inline\-toc\.[\r\n\s]*^([\s\S]*?)^\s*$/m;

// TODO: split this into more focused files.
const parseNanocPartials = contentString => contentString.split(/<%=\s+partial\s+partial_version\('([^')]*)'\)[^%>]*%>/);
const tryRetrieveInlineTOC = contentString => contentString.match(INLINE_TOC_REGEX);
const removeInlineTOC = contentString => contentString.replace(INLINE_TOC_REGEX, '');
const processTOCItems = tocItem => {
  if(isEmpty(tocItem)) {
    return tocItem;
  }
  if(Array.isArray(tocItem)) {
    return tocItem.map(processTOCItems);
  }
  if(isPlainObject(tocItem)) {
    return {
      content: Object.entries(tocItem).map(([ key, values ]) => ({ key, values: processTOCItems(values) })),
    };
  }
  const linkParts = tocItem.match(/(.+)#(.+)/);
  const linkTitle = (linkParts && linkParts[1]) || tocItem;
  const link = (linkParts && linkParts[2]) || lowerFirst(tocItem);
  return {
    linkTitle,
    link: link.replace(/\s/g, '-')
  }
}
const retrieveAndReplaceInlineTOC = contentString => ({
  noInlineTOC: removeInlineTOC(contentString),
  inlineTOCOnly: (tryRetrieveInlineTOC(contentString) ?? [,null])[1]
});

const removeFalsy = dataArray => dataArray.filter(identity);
const makeTypeFromParentType = type => node => upperFirst(camelCase(`${node.internal.type}${type}`));

const createNodesFromPath = (type, { createNode, createNodeId, createContentDigest }) => path => {
  const pieces = path.split('/');
  const values = {
    link: '',
    label: '',
    level: ROOT_LEVEL
  };
  const breadcrumbs = pieces.map((piece, i) => {
    values.level = values.level + i > MAX_LEVEL ? MAX_LEVEL : values.level + i;
    values.link = `${values.link}/${piece}`;
    values.label = upperFirst(piece);
    values.id = createNodeId(`${values.link} >>> Path`)

    return { ...values };
  });
  breadcrumbs.forEach((breadcrumb, i) => {
    const newNodeData = {
      ...breadcrumb,
      children: [],
    };
    if(breadcrumbs[i-1]) {
      newNodeData.parent = breadcrumbs[i-1].id;
    }
    const newNodeInternals = {
      contentDigest: createContentDigest(breadcrumb.link),
      type,
      mediaType: 'text/plain'
    };
    const pathNode = {
      ...newNodeData,
      internal: newNodeInternals,
    }
    createNode(pathNode);
  });
}

const constructDataObjectsFromStrings = (contentStrings, frontmatterMeta) => {
  contentStrings = frontmatterMeta !== NO_MATCH && frontmatterMeta.attributes ?
    contentStrings.map(content => enhancedParse(content, frontmatterMeta.attributes)) :
    contentStrings;
  const dataObjects = contentStrings.map(
    (data, i) => i % 2 === 0 ?
      { data: data, type: DataTypes.Html } :
      { data: data, type: DataTypes.Partial }
  );
  return dataObjects;
}

const flattenContentOrderedList = contentOrderedList => contentOrderedList.reduce((acc, {data, type}) => {
  if(Array.isArray(data)) {
      return acc.concat(flattenContentOrderedList(data));
  }
  return acc.concat([{ data, type }]);
}, []);

// Source: https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/creating-a-transformer-plugin/
const transformNanocTextiles = (node, content, id, type, { createNodesFromPath, createContentDigest, createNodeId }) => updateWithTransform => {
    const slug = node.relativePath.replace(/(.*)\.[^.]+$/, "$1");
    const preTextileTransform = preParser(content);
    // We could re-arrange & limit this to the last array item if we are confident that no partials will appear in the API reference.
    const {
      noInlineTOC,
      inlineTOCOnly
    } = retrieveAndReplaceInlineTOC(preTextileTransform);
    const loadedInlineTOC = yaml.load(inlineTOCOnly,'utf-8');
    const inlineTOCLinks = {
      tableOfContents: processTOCItems(loadedInlineTOC)
    };
    const inlineTOCNode = merge(inlineTOCLinks, {
      id: createNodeId(`${id} >>> InlineTOC`),
      children: [],
      parent: node.id,
      slug,
      internal: {
        contentDigest: createContentDigest(inlineTOCLinks),
        type: makeTypeFromParentType('InlineTOC')(node),
      }
    });
    updateWithTransform({ parent: node, child: inlineTOCNode });
    // if we need it, remember to use DOMParser not Cheerio!
    const withPartials = parseNanocPartials(noInlineTOC);

    const withoutFalsyValues = removeFalsy(withPartials);

    const frontmatterMeta = tryRetrieveMetaData(withoutFalsyValues[0]);
    if(frontmatterMeta !== NO_MATCH) {
      withoutFalsyValues[0] = frontmatterMeta.body;
    }

    const asDataObjects = constructDataObjectsFromStrings(withoutFalsyValues, frontmatterMeta);
    const newNodeData = {
      contentOrderedList: asDataObjects,
      id,
      children: [],
      parent: node.id,
    };

    if(frontmatterMeta !== NO_MATCH) {
      newNodeData.meta = filterAllowedMetaFields(frontmatterMeta.attributes);
    }

    const newNodeInternals = {
      contentDigest: createContentDigest(asDataObjects),
      type,
      mediaType: 'text/html'
    };
    if(node.sourceInstanceName === 'textile-partials') {
      newNodeInternals.type = `${type}Partial`;
      newNodeData.relativePath = node.relativePath;
    } else {
      createNodesFromPath(node.relativePath.replace(/\.[^/.]+$/, ""));
      // Partials should never have a slug, every other page type needs one.
      newNodeData.slug = slug;
    }
    const htmlNode = {
      ...newNodeData,
      internal: newNodeInternals,
    }
    if (content.id) {
        htmlNode[`htmlId`] = content.id
    }
    updateWithTransform({ parent: node, child: htmlNode })
}

const maybeRetrievePartial = graphql => async ({ data, type }) => {
  if(type !== DataTypes.Partial) {
    return { data, type };
  }
  const result = await graphql(`
      query {
        fileHtmlPartial(relativePath: { eq:"${data}.textile"}) {
          contentOrderedList {
            data
            type
          }
        }
      }
    `)
  const partial = result.data.fileHtmlPartial;
  const retrievePartialFromGraphQL = maybeRetrievePartial(graphql);
  let contentOrderedList = partial && partial.contentOrderedList;
  if(partial) {
    contentOrderedList = await Promise.all(contentOrderedList.map(retrievePartialFromGraphQL));
  }

  return {
    data: contentOrderedList,
    type
  };
}

module.exports = {
  createNodesFromPath,
  transformNanocTextiles,
  makeTypeFromParentType,
  maybeRetrievePartial,
  flattenContentOrderedList
};