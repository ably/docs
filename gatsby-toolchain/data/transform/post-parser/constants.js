// Shared constants between different post-parse operations
/**
 * Any number of opening or closing XML tags, inc. HTML tags, followed by any number of spaces
 */
const TAG_REGEX_STRING = '(?:<[\\w \\/]+>)*[\\s\\r\\n]*';

module.exports = {
    TAG_REGEX_STRING
}