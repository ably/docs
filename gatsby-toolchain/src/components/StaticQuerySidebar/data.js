import { groupBy } from 'lodash';
import { MAX_LEVEL, ROOT_LEVEL } from '../Sidebar/consts';

const sidebarDataFromPageFurniture = data => data.map(({ label, link, level = ROOT_LEVEL, contentString = false, contentArray = false }) => {
    const result = {
        label,
        link,
        level,
        content: contentArray ? sidebarDataFromPageFurniture(contentArray) : contentString
    }
    return result;
})

const range = (start, end) => Array(end - start).fill().map((_, i) => i + start);

const sidebarDataFromDocumentPaths = data => {
    const objects = data.map(({ node: { id, label, level, link, parent = { id: false } }}) => ({ id, label, level, link, parent }));
    const objectsByLevel = groupBy(objects, object => object.level);
    // Avoids possible GC issues caused by setting a variable with let. Premature optimization but a simple one.
    range(ROOT_LEVEL, MAX_LEVEL).forEach(currentLevel => {
        objectsByLevel[currentLevel] && objectsByLevel[currentLevel].map(object => {
            object.content = objectsByLevel[currentLevel+1] ?
                [{...object, label: 'Overview' }].concat(
                    objectsByLevel[currentLevel+1]
                        .filter(({ parent: { id } }) => id === object.id)
                ) :
                null;
            return object;
        });
    })
    return objectsByLevel[ROOT_LEVEL];
}

export { sidebarDataFromPageFurniture, sidebarDataFromDocumentPaths };