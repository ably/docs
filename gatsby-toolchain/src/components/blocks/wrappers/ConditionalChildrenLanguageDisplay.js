import React, { Children, useContext } from 'react';
import { DEFAULT_LANGUAGE, IGNORED_LANGUAGES_FOR_DISPLAY } from '../../../../data/createPages/createPageVariants';
import PageLanguageContext from '../../../contexts/page-language-context';

const makeGroup = (lang, index) => ({
    start: index,
    end: index,
    primary: lang,
    languages: [lang]
});

const assignPrimary = (group, lang, targetLanguage) => {
    group.languages.push(lang);
    if(lang === targetLanguage) {
        return {
            ...group,
            primary: targetLanguage
        }
    }
    if(lang === DEFAULT_LANGUAGE && group.primary !== targetLanguage) {
        return {
            ...group,
            primary: DEFAULT_LANGUAGE
        }
    }
    return group;
}

const addToFilter = (group, toFilter) => {
    for(let i = 0; group.start + i <= group.end; ++i) {
        toFilter[group.start+i] = group.languages[i] !== group.primary;
    }
}

const ConditionalChildrenLanguageDisplay = ({ children }) => {
    const language = useContext(PageLanguageContext);
    let currentGroup = false;
    const childLanguageGroups = [];
    const toFilter = [];
    Children.forEach(
        children,
        ({ props: { attribs = null } }, index) => {
            if(
                attribs &&
                attribs.lang &&
                !IGNORED_LANGUAGES_FOR_DISPLAY.includes(attribs.lang)
            ) {
                if(!currentGroup) {
                    currentGroup = makeGroup(attribs.lang, index);
                } else if(currentGroup.languages.includes(attribs.lang)) {
                    // A repeated lang is possibly user error, but if it is not it would be hard to work out why your documentation doesn't show.
                    addToFilter(currentGroup, toFilter);
                    childLanguageGroups.push({...currentGroup});
                    currentGroup = makeGroup(attribs.lang, index);
                } else {
                    currentGroup.end = index;
                    currentGroup = assignPrimary(currentGroup, attribs.lang, language);
                }
            } else {
                if(!!currentGroup) {
                    addToFilter(currentGroup, toFilter);
                    childLanguageGroups.push({...currentGroup});
                }
                currentGroup = false;
                toFilter[index] = false;
            }
        }
    );
    if(!!currentGroup) {
        addToFilter(currentGroup, toFilter);
        childLanguageGroups.push({...currentGroup});
    }
    return Children.map(
        children,
        (child, index) => {
            if(toFilter[index]) {
                return null;
            }
            return child;
        }
    );
};

export default ConditionalChildrenLanguageDisplay;