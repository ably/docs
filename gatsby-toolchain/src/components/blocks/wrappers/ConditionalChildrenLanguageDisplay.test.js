import React from 'react';
import TestRenderer from 'react-test-renderer';
import cheerio from 'cheerio';
import { DEFAULT_LANGUAGE } from '../../../../data/createPages/constants';
import textile from 'textile-js';
import { htmlParser, liftLangAttributes } from '../../../../data/html-parser';
import { postParser } from '../../../../data/transform/post-parser';
import { preParser } from '../../../../data/transform/pre-parser';
import ConditionalChildrenLanguageDisplay from './ConditionalChildrenLanguageDisplay';
import blocksFromData from '../blocks-from-data';

const rawData = `
- <div lang="jsall">callback</div> := is a function of the form @function(err)@ which is called upon completion
- <div lang="java">listener</div> := Listener to be notified on completion<br>__Type: "@CompletionListener@":#completion-listener__
- <div lang="csharp">callback</div> := is an action of the form @Action<bool, ErrorInfo>@ which is called upon completion
- <div lang="ruby">&block</div> := yields upon successfully publishing the message
- <div lang="objc,swift">callback</div> := called upon publishing the message, or with an error
`;

const twoChildrenInstance = TestRenderer.create(<ConditionalChildrenLanguageDisplay>
    {blocksFromData([{
        'name': 'span',
        'data': 'content',
        'attribs': { lang: DEFAULT_LANGUAGE }
    }, {
        'name': 'span',
        'data': 'alternative',
        'attribs': { lang: 'javascript' }
    }])}
</ConditionalChildrenLanguageDisplay>).root;

const sameLanguageTwoChildrenInstance = TestRenderer.create(<ConditionalChildrenLanguageDisplay>
    {blocksFromData([{
        'name': 'span',
        'data': 'content',
        'attribs': { lang: 'javascript' }
    }, {
        'name': 'span',
        'data': 'alternative',
        'attribs': { lang: 'javascript' }
    }])}
</ConditionalChildrenLanguageDisplay>).root;

describe('ConditionalChildrenLanguageDisplay only displays one child of alternatives', () => {
    it('A basic instance of two children with different lang attributes only shows the default language option', () => {
        expect(twoChildrenInstance.findAllByType('span').length).toBe(1);
    });

    it('Two different instances in a row shows both language options', () => {
        expect(sameLanguageTwoChildrenInstance.findAllByType('span').length).toBe(2);
    });
});

describe('Integration: ConditionalChildrenLanguageDisplay only displays one <dt><dd> pair of children from alternatives for parsed definition lists', () => {
    const parsedData = postParser(textile(preParser(rawData)));
    it('Text parser stage returns expected results', () => {
        expect(parsedData.replace(/\s/g,'')).toEqual(`<dl>
            <dt><div lang=\"javascript,nodejs\">callback</div></dt>
            <dd>is a function of the form <code>function(err)</code> which is called upon completion</dd>
            <dt><div lang=\"java\">listener</div></dt>
            <dd>Listener to be notified on completion<br /><i>Type: <a href=\"#completion-listener\"><code>CompletionListener</code></a></i></dd>
            <dt><div lang=\"csharp\">callback</div></dt>
            <dd>is an action of the form <code>Action&lt;bool, ErrorInfo&gt;</code> which is called upon completion</dd>
            <dt><div lang=\"ruby\">&amp;block</div></dt>
            <dd>yields upon successfully publishing the message</dd>
            <dt><div lang=\"objc,swift\">callback</div></dt>
            <dd>called upon publishing the message, or with an error</dd>
        </dl>`.replace(/\s/g,''));
    });

    // cf. lift-language-attributes.test.js
    // Included here to allow for easy debugging of which particular stage of the process is broken
    it('Lift language attributes stage returns expected results', () => {
        const loadedDom = cheerio.load(parsedData, null);
        liftLangAttributes(loadedDom);
        expect(loadedDom('body *').html().replace(/\s/g,'')).toEqual(`<div lang=\"javascript,nodejs\"><dt><div>callback</div></dt><dd>is a function of the form <code>function(err)</code> which is called upon completion</dd></div>
    	<div lang=\"java\"><dt><div>listener</div></dt><dd>Listener to be notified on completion<br><i>Type: <a href=\"#completion-listener\"><code>CompletionListener</code></a></i></dd></div>
    	<div lang=\"csharp\"><dt><div>callback</div></dt><dd>is an action of the form <code>Action&lt;bool, ErrorInfo&gt;</code> which is called upon completion</dd></div>
    	<div lang=\"ruby\"><dt><div>&amp;block</div></dt><dd>yields upon successfully publishing the message</dd></div>
    	<div lang=\"objc,swift\"><dt><div>callback</div></dt><dd>called upon publishing the message, or with an error</dd></div>
`.replace(/\s/g,''));
    });

    const htmlParsed = htmlParser(parsedData);
    it('HTML parser stage returns expected results', () => {    
        expect(htmlParsed[0].data[0].data).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    "attribs": {"lang": "nodejs"},
                    "name": "div",
                    "data": expect.arrayContaining([
                        expect.objectContaining({
                            "name": "dt"
                        }),
                        expect.objectContaining({
                            "name": "dd"
                        })
                    ])
                })
            ])
        );
    });

    it('ConditionalChildrenLanguageDisplay displays the expected results from HTML data', () => {
        const renderedData = TestRenderer.create(<ConditionalChildrenLanguageDisplay>
            {blocksFromData(htmlParsed)}
        </ConditionalChildrenLanguageDisplay>);
        const tree = renderedData.toJSON();
        expect(tree).toMatchSnapshot();
        expect(renderedData.root.findAllByType('dt').length).toBe(1);
        expect(renderedData.root.findAllByType('dd').length).toBe(1);
    });
});