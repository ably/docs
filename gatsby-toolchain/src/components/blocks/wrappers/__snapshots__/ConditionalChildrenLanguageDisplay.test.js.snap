// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: ConditionalChildrenLanguageDisplay only displays one <dt><dd> pair of children from alternatives for parsed definition lists ConditionalChildrenLanguageDisplay displays the expected results from HTML data 1`] = `
<dl>
  
	
  <div
    lang="nodejs"
  >
    <dt>
      <div>
        callback
      </div>
    </dt>
    <dd>
      is a function of the form 
      <code
        className="sc-jrQzAO kSxfLf"
      >
        function(err)
      </code>
       which is called upon completion
    </dd>
  </div>
  
	
  

</dl>
`;
